corehq/ex-submodules/couchforms/models.py

```diff
@@ -346,27 +346,32 @@ def _meta_to_json(meta):
     def xml_md5(self):
         return hashlib.md5(self.get_xml().encode('utf-8')).hexdigest()
 
-    def archive(self, user_id=None):
-        if self.is_archived:
+    def archive(self, user_id=None, retry_archive=False):
+        if self.is_archived and not retry_archive:
             return
-        self.doc_type = "XFormArchived"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='archive',
-        ))
-        self.save()
-        xform_archived.send(sender="couchforms", xform=self)
 
-    def unarchive(self, user_id=None):
-        if not self.is_archived:
+        from corehq.form_processor.submission_post import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True):
+            self.doc_type = "XFormArchived"
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='archive',
+            ))
+            self.save()
+            xform_archived.send(sender="couchforms", xform=self)
+
+    def unarchive(self, user_id=None, retry_archive=False):
+        if not self.is_archived and not retry_archive:
             return
-        self.doc_type = "XFormInstance"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='unarchive',
-        ))
-        XFormInstance.save(self)  # subclasses explicitly set the doc type so force regular save
-        xform_unarchived.send(sender="couchforms", xform=self)
+        from corehq.form_processor.submission_post import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False):
+            self.doc_type = "XFormInstance"
+            self.history.append(XFormOperation(
```

Comment:
Do we care if we end up with duplicate operations? Would be nice if we could avoid it.

Reply:
I changed it so that the history only gets updated at the end, after the form is archived and the case is rebuilt: https://github.com/dimagi/commcare-hq/pull/22470/commits/49d769180a6f9dc7fb3a779b14bb2181499ae2a1

One problem with this solution is that a form could be archived, but not have the archive show up in its history for the 5 minutes while it is being reprocessed.

On another branch, I worked on another solution where the history is only updated once, and not updated on retries, which has the reverse problem of showing that a form was archived before the case is rebuilt. https://github.com/dimagi/commcare-hq/commit/30fd6d86a3cdcc01adaf4fd388782b98e7cfc39a

I like the solution that I did in this branch better, it could lead to some confusion, but seems to target the root issue that was confusing to users in this ticket.

Reply:
This won't work since rebuilding the case relies on the form being marked as 'archived' otherwise the case rebuild will have no effect. I think it's fine to do the alternate approach in https://github.com/dimagi/commcare-hq/commit/30fd6d86a3cdcc01adaf4fd388782b98e7cfc39a

Reply:
Hm, the tests do pass, which checks if the case is rebuilt properly. Are you referring to `is_archived`? Because that has to do with checking `self.doc_type`, which does get updated before the case is rebuilt

Reply:
Did the other approach here: https://github.com/dimagi/commcare-hq/pull/22470/commits/6e277f171ae513fec02b658970e11cf07bbda346

Reply:
https://github.com/dimagi/commcare-hq/pull/22470#issuecomment-441293835